File: creational/typescript/factory_method_structured/concrete_products.ts
import { Document, PaymentProcessor } from './interfaces';

export class TextDocument implements Document {
  open(): void { console.log('Opening Text Document in text editor...'); }
  close(): void { console.log('Closing Text Document...'); }
  save(): void { console.log('Saving Text Document as .txt file...'); }
  getType(): string { return 'Text Document'; }
}

export class PDFDocument implements Document {
  open(): void { console.log('Opening PDF Document in PDF viewer...'); }
  close(): void { console.log('Closing PDF Document...'); }
  save(): void { console.log('Saving PDF Document as .pdf file...'); }
  getType(): string { return 'PDF Document'; }
}

export class HtmlDocument implements Document {
  open(): void { console.log('Opening HTML Document in web browser...'); }
  close(): void { console.log('Closing HTML Document...'); }
  save(): void { console.log('Saving HTML Document as .html file...'); }
  getType(): string { return 'HTML Document'; }
}

export class CreditCardProcessor implements PaymentProcessor {
  processPayment(amount: number): void { console.log(`Processing credit card payment of $${amount}`); }
  refundPayment(amount: number): void { console.log(`Refunding credit card payment of $${amount}`); }
}

export class PayPalProcessor implements PaymentProcessor {
  processPayment(amount: number): void { console.log(`Processing PayPal payment of $${amount}`); }
  refundPayment(amount: number): void { console.log(`Refunding PayPal payment of $${amount}`); }
}

export class CryptoProcessor implements PaymentProcessor {
  processPayment(amount: number): void { console.log(`Processing cryptocurrency payment of $${amount}`); }
  refundPayment(amount: number): void { console.log(`Refunding cryptocurrency payment of $${amount}`); }
}

File: creational/dart/factory_method_structured/concrete_products.dart
import 'interfaces.dart';

class TextDocument extends Document {
  @override
  void open() => print('Opening Text Document in text editor...');
  @override
  void close() => print('Closing Text Document...');
  @override
  void save() => print('Saving Text Document as .txt file...');
  @override
  String get type => 'Text Document';
}

class PDFDocument extends Document {
  @override
  void open() => print('Opening PDF Document in PDF viewer...');
  @override
  void close() => print('Closing PDF Document...');
  @override
  void save() => print('Saving PDF Document as .pdf file...');
  @override
  String get type => 'PDF Document';
}

class HtmlDocument extends Document {
  @override
  void open() => print('Opening HTML Document in web browser...');
  @override
  void close() => print('Closing HTML Document...');
  @override
  void save() => print('Saving HTML Document as .html file...');
  @override
  String get type => 'HTML Document';
}

class CreditCardProcessor extends PaymentProcessor {
  @override
  void processPayment(int amount) => print('Processing credit card payment of \$$amount');
  @override
  void refundPayment(int amount) => print('Refunding credit card payment of \$$amount');
}

class PayPalProcessor extends PaymentProcessor {
  @override
  void processPayment(int amount) => print('Processing PayPal payment of \$$amount');
  @override
  void refundPayment(int amount) => print('Refunding PayPal payment of \$$amount');
}

class CryptoProcessor extends PaymentProcessor {
  @override
  void processPayment(int amount) => print('Processing cryptocurrency payment of \$$amount');
  @override
  void refundPayment(int amount) => print('Refunding cryptocurrency payment of \$$amount');
}
